modules/gateways/btcpay.php

```diff
@@ -96,12 +100,22 @@ function btcpay_link($params)
     $country   = $params['clientdetails']['country'];
     $phone     = $params['clientdetails']['phonenumber'];
 
+    // Tor support
+    $parsedurl = parse_url($params['systemurl']);
+    $is_tor_enabled = preg_match("/\.onion$/", $_SERVER['HTTP_HOST']) && $params['btcpayUrlTor'] != '';
+    if (is_array($parsedurl) && $is_tor_enabled) {
+	$systemtor = "http://{$_SERVER['HTTP_HOST']}" . $parsedurl['path'];
+    } else {
+        $systemtor = "";
+    }
+
     // System Variables
-    $systemurl = $params['systemurl'];
+    $systemurl = $systemtor != '' ? $systemtor : $params['systemurl'];
 
     $post = array(
         'invoiceId'     => $invoiceid,
         'systemURL'     => $systemurl,
+        'ipnURL'        => $params['systemurl'] . '/modules/gateways/callback/btcpay.php',
```

Comment:
Ok I see you add the ipn as separate param here. Any reason you do not use `$systemurl` here?

Reply:
@ndeet, because I have `$systemurl = $systemtor != '' ? $systemtor : $params['systemurl'];`
IPN is server-side thus I don't need to use .onion when client does

Reply:
Ok but it will break ipn if you run WHMCS (and BTCPay) only on tor without clearnet?

Reply:
> Ok but it will break ipn if you run WHMCS (and BTCPay) only on tor without clearnet?

In that case `$params['systemurl']` is already .onion so I doubt that

Reply:
Right, that works. :+1: